mscore/pagesettings.cpp

```diff
@@ -108,159 +105,217 @@ void PageSettings::setScore(Score* s)
-//   mmClicked
+//   typeChanged - populates sizesList based on typesList
 //---------------------------------------------------------
 
-void PageSettings::mmClicked()
+void PageSettings::typeChanged(int idx)
       {
-      mmUnit = true;
-      updateValues();
+      switch (PaperType(idx)) {
+            case PaperType::Metric:
+                  _sizes = &MScore::sizesMetric;
+                  break;
+            case PaperType::Imperial:
+                  _sizes = &MScore::sizesImperial;
+                  break;
+            case PaperType::Other:
+                  _sizes = &MScore::sizesOther;
+                  break;
+            default:
+                  Q_ASSERT_X(false, "PageSettings::typeChanged", "PaperType not set!");
+                  return;
+            }
+      sizesList->clear(); // Custom is at the top of every list
+      sizesList->addItem(QPageSize::name(QPageSize::Custom), int(QPageSize::Custom));
+      for (std::set<int>::iterator i = _sizes->begin(); i != _sizes->end(); ++i)
+            sizesList->addItem(QPageSize::name(QPageSize::PageSizeId(*i)), *i);
       }
 
 //---------------------------------------------------------
-//   orientationClicked
-//    swap width/height
+//   setPageSize - a helper for page size, width, height changes
 //---------------------------------------------------------
 
-void PageSettings::orientationClicked()
+void PageSettings::setPageSize(QPageSize::PageSizeId psid)
       {
-      qreal w = preview->score()->styleD(Sid::pageWidth);
-      qreal h = preview->score()->styleD(Sid::pageHeight);
+      Score* score = preview->score();
+      MStyle& style = score->style();
+      QPageSize::Unit unit = QPageSize::Unit(style.pageOdd().units());
+      QPageSize qps;
 
-      preview->score()->style().set(Sid::pageWidth, h);
-      preview->score()->style().set(Sid::pageHeight, w);
+      bool isPreset = (psid != QPageSize::Custom);
+      if (isPreset)
+            qps = QPageSize(psid);
+      else
+            qps = QPageSize(QSizeF(pageWidth->value(), pageHeight->value()),
+                  unit,
+                  QPageSize::name(psid),
+                  QPageSize::ExactMatch);
+
+      style.setPageSize(qps);
+      style.pageOdd ().setPageSize(qps);
+      style.pageEven().setPageSize(qps);
+
+      if (isPreset) {
+            QRectF rect = qps.rect(unit);
+            updateWidthHeight(rect);
+            updatePreview();
+            }
+      }
+//---------------------------------------------------------
+//   sizeChanged
+//---------------------------------------------------------
 
-      double f = mmUnit ? 1.0/INCH : 1.0;
-      preview->score()->style().set(Sid::pagePrintableWidth, h - (oddPageLeftMargin->value() + oddPageRightMargin->value()) * f);
-      updateValues();
-      updatePreview(0);
+void PageSettings::sizeChanged(int idx)
+      {
+      if (idx >= 0) // should never be -1, but just in case...
```

Comment:
add `Q_ASSERT()` to make sure the value is never negative.

Reply:
OK, no problem.